File: tests/verify_files.cpp
// verify_files.cpp
// Compares two files by computing and comparing their CIDs

#include <iostream>
#include <string>
#include "FileUtils.h"
#include "HashUtils.h"

using namespace std;

string computeCID(const string &filepath)
{
    // Chunk the file
    vector<Chunk> chunks = SplitFileIntoChunks(filepath);
    if (chunks.empty())
    {
        return "";
    }

    // Hash all chunks
    HashAllChunks(chunks);

    // Collect chunk hashes
    vector<string> chunkHashes;
    for (const Chunk &chunk : chunks)
    {
        chunkHashes.push_back(chunk.hash);
    }

    // Compute root hash (CID)
    return ComputeRootHash(chunkHashes);
}

int main(int argc, char *argv[])
{
    if (argc != 3)
    {
        cerr << "Usage: " << argv[0] << " <original_file> <reconstructed_file>" << endl;
        return 1;
    }

    string originalPath = argv[1];
    string reconstructedPath = argv[2];

    cout << "Computing CID for original file..." << endl;
    string originalCID = computeCID(originalPath);
    if (originalCID.empty())
    {
        cerr << "Error: Could not process original file" << endl;
        return 1;
    }

    cout << "Computing CID for reconstructed file..." << endl;
    string reconstructedCID = computeCID(reconstructedPath);
    if (reconstructedCID.empty())
    {
        cerr << "Error: Could not process reconstructed file" << endl;
        return 1;
    }

    cout << "\n--- Results ---" << endl;
    cout << "Original CID:      " << originalCID << endl;
    cout << "Reconstructed CID: " << reconstructedCID << endl;

    if (originalCID == reconstructedCID)
    {
        cout << "\nVERIFIED: Files are identical" << endl;
        return 0;
    }
    else
    {
        cout << "\nMISMATCH: Files differ" << endl;
        return 1;
    }
}

File: src/main/java/com/distributed/storage/VerifyFiles.java
package com.distributed.storage;

import java.util.ArrayList;
import java.util.List;

public class VerifyFiles {

    public static String computeCID(String filepath) {
        // Chunk the file
        List<Chunk> chunks = FileUtils.splitFileIntoChunks(filepath);
        if (chunks.isEmpty()) {
            return "";
        }

        // Hash all chunks
        HashUtils.hashAllChunks(chunks);

        // Collect chunk hashes
        List<String> chunkHashes = new ArrayList<>();
        for (Chunk chunk : chunks) {
            chunkHashes.add(chunk.hash);
        }

        // Compute root hash (CID)
        return HashUtils.computeRootHash(chunkHashes);
    }

    public static void main(String[] args) {
        if (args.length != 2) {
            System.err.println("Usage: java VerifyFiles <original_file> <reconstructed_file>");
            System.exit(1);
        }

        String originalPath = args[0];
        String reconstructedPath = args[1];

        System.out.println("Computing CID for original file...");
        String originalCID = computeCID(originalPath);
        if (originalCID.isEmpty()) {
            System.err.println("Error: Could not process original file");
            System.exit(1);
        }

        System.out.println("Computing CID for reconstructed file...");
        String reconstructedCID = computeCID(reconstructedPath);
        if (reconstructedCID.isEmpty()) {
            System.err.println("Error: Could not process reconstructed file");
            System.exit(1);
        }

        System.out.println("\n--- Results ---");
        System.out.println("Original CID:      " + originalCID);
        System.out.println("Reconstructed CID: " + reconstructedCID);

        if (originalCID.equals(reconstructedCID)) {
            System.out.println("\nVERIFIED: Files are identical");
            System.exit(0);
        } else {
            System.out.println("\nMISMATCH: Files differ");
            System.exit(1);
        }
    }
}
